docs/config.md
# DefraDB configuration (YAML)

The default DefraDB directory is `$HOME/.defradb`. It can be changed via the --rootdir CLI flag.

Relative paths are interpreted as being rooted in the DefraDB directory.

## `development`

Enables a set of features that make development easier but should not be enabled in production.

## `datastore.store`

Store can be badger or memory. Defaults to `badger`.

- badger: fast pure Go key-value store optimized for SSDs (https://github.com/dgraph-io/badger)
- memory: in-memory version of badger

## `datastore.maxtxnretries`

The number of retries to make in the event of a transaction conflict. Defaults to `5`.

Currently this is only used within the P2P system and will not affect operations initiated by users.

## `datastore.noencryption`

Skip generating an encryption key. Encryption at rest will be disabled. **WARNING**: This cannot be undone.

## `datastore.badger.path`

The path to the database data file(s). Defaults to `data`.

## `datastore.badger.valuelogfilesize`

Maximum file size of the value log files.

## `api.address`

Address of the HTTP API to listen on or connect to. Defaults to `127.0.0.1:9181`.

## `api.allowed-origins`

The list of origins a cross-domain request can be executed from.

## `api.pubkeypath`

The path to the public key file for TLS / HTTPS.

## `api.privkeypath`

The path to the private key file for TLS / HTTPS.

## `net.p2pdisabled`

Whether P2P networking is disabled. Defaults to `false`.

## `net.p2paddresses`

List of addresses for the P2P network to listen on. Defaults to `/ip4/127.0.0.1/tcp/9171`.

## `net.pubsubenabled`

Whether PubSub is enabled. Defaults to `true`.

## `net.peers`

List of peers to boostrap with, specified as multiaddresses.

https://docs.libp2p.io/concepts/addressing/

## `net.relay`

Enable libp2p's Circuit relay transport protocol. Defaults to `false`.

https://docs.libp2p.io/concepts/circuit-relay/

## `net.wifiawareenabled`

Enable Wi-Fi Aware local mesh peer discovery (mDNS-backed). Defaults to `false`.

Nodes on the same local network discover and connect to each other
automatically, without configuring `net.peers`. To be reachable across
machines the node must listen on a non-loopback address, e.g.
`net.p2paddresses: [/ip4/0.0.0.0/tcp/9171]`, and the network must permit
multicast between clients.

## `log.level`

Log level to use. Options are `info` or `error`. Defaults to `info`.

## `log.output`

Log output path. Options are `stderr` or `stdout`. Defaults to `stderr`.

## `log.format`

Log format to use. Options are `text` or `json`. Defaults to `text`.

## `log.stacktrace`

Include stacktrace in error and fatal logs. Defaults to `false`.

## `log.source`

Include source location in logs. Defaults to `false`.

## `log.overrides`

Logger config overrides. Format `<name>,<key>=<val>,...;<name>,...`.

## `log.colordisabled`

Disable colored log output. Defaults to `false`.

## `keyring.path`

Path to store encrypted key files in. Defaults to `keys`.

## `keyring.disabled`

Disable the keyring and generate ephemeral keys instead. Defaults to `false`.

## `keyring.namespace`

The service name to use when using the system keyring. Defaults to `defradb`.

## `keyring.backend`

Keyring backend to use. Defaults to `file`.

- `file` Stores keys in encrypted files
- `system` Stores keys in the OS managed keyring

## `lens.runtime`

The LensVM wasm runtime to run lens modules in.

Possible values:
- `wasm-time` (default): https://github.com/bytecodealliance/wasmtime-go
- `wazero`: https://github.com/tetratelabs/wazero

## `acp.document.type`

The type of Document ACP to use.

Possible values:
- `local` (default): local-only Document ACP
- `source-hub` source hub Document ACP: https://github.com/sourcenetwork/sourcehub

## `acp.document.sourceHub.ChainID`

The ID of the SourceHub chain to store ACP data in. Required when using `acp.document.type`:`source-hub`.

## `acp.document.sourceHub.GRPCAddress`

The address of the SourceHub GRPC server. Required when using `acp.document.type`:`source-hub`.

## `acp.document.sourceHub.CometRPCAddress`

The address of the SourceHub Comet RPC server. Required when using `acp.document.type`:`source-hub`.

## `acp.document.sourceHub.KeyName`

The name of the key in the keyring where the SourceHub credentials used to sign (and pay for) SourceHub
transactions created by the node is stored. Required when using `acp.document.type`:`source-hub`.

## `acp.document.sourceHub.address`

The SourceHub address of the actor that client-side actions should permit to make SourceHub actions on
their behalf.  This is a client-side only config param.  It is required if the client wishes to make
SourceHub ACP requests in order to create protected data.

## `secretfile`

Path to the file containing secrets. Defaults to `.env`.

## `telemetry.disabled`

Disables telemetry reporting. Telemetry is only enabled in builds that use the `telemetry` flag. Defaults to `false`.


cli/config/config.go
// Copyright 2024 Democratized Data Foundation
//
// Use of this software is governed by the Business Source License
// included in the file licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0, included in the file
// licenses/APL.txt.

package config

import (
	"errors"
	"os"
	"path/filepath"
	"strings"

	"github.com/joho/godotenv"
	"github.com/spf13/pflag"
	"github.com/spf13/viper"

	"github.com/sourcenetwork/corelog"
)

const (
	ConfigStoreBadger   = "badger"
	ConfigStoreMemory   = "memory"
	ConfigLogFormatJSON = "json"
	ConfigLogFormatCSV  = "csv"
	ConfigLogLevelInfo  = "info"
	ConfigLogLevelDebug = "debug"
	ConfigLogLevelError = "error"
	ConfigLogLevelFatal = "fatal"
)

// ConfigPaths are config keys that will be made relative to the rootdir
var ConfigPaths = []string{
	"datastore.badger.path",
	"api.pubkeypath",
	"api.privkeypath",
	"keyring.path",
}

// ConfigFlags is a mapping of cli flag names to config keys to bind.
var ConfigFlags = map[string]string{
	"log-level":                  "log.level",
	"log-output":                 "log.output",
	"log-format":                 "log.format",
	"log-stacktrace":             "log.stacktrace",
	"log-source":                 "log.source",
	"log-overrides":              "log.overrides",
	"no-log-color":               "log.colordisabled",
	"url":                        "api.address",
	"max-txn-retries":            "datastore.maxtxnretries",
	"store":                      "datastore.store",
	"no-encryption":              "datastore.noencryption",
	"no-searchable-encryption":   "datastore.nosearchableencryption",
	"no-signing":                 "datastore.nosigning",
	"default-key-type":           "datastore.defaultkeytype",
	"valuelogfilesize":           "datastore.badger.valuelogfilesize",
	"peers":                      "net.peers",
	"p2paddr":                    "net.p2paddresses",
	"no-p2p":                     "net.p2pdisabled",
	"relay":                      "net.relay",
	"enable-wifi-aware":          "net.wifiawareenabled",
	"allowed-origins":            "api.allowed-origins",
	"pubkeypath":                 "api.pubkeypath",
	"privkeypath":                "api.privkeypath",
	"keyring-namespace":          "keyring.namespace",
	"keyring-backend":            "keyring.backend",
	"keyring-path":               "keyring.path",
	"no-keyring":                 "keyring.disabled",
	"document-acp-type":          "acp.document.type",
	"source-hub-address":         "acp.document.sourceHub.address",
	"development":                "development",
	"secret-file":                "secretfile",
	"no-telemetry":               "telemetry.disabled",
	"replicator-retry-intervals": "replicator.retryintervals",
}

// ConfigDefaults contains default values for config entries.
var ConfigDefaults = map[string]any{
	"api.address":                       "127.0.0.1:9181",
	"api.allowed-origins":               []string{},
	"datastore.badger.path":             "data",
	"datastore.maxtxnretries":           5,
	"datastore.store":                   "badger",
	"datastore.badger.valuelogfilesize": 1 << 30,
	"development":                       false,
	"net.p2pdisabled":                   false,
	"net.p2paddresses":                  []string{"/ip4/127.0.0.1/tcp/9171"},
	"net.peers":                         []string{},
	"net.pubSubEnabled":                 true,
	"net.relay":                         false,
	"net.wifiawareenabled":              false,
	"keyring.backend":                   "file",
	"keyring.disabled":                  false,
	"keyring.namespace":                 "defradb",
	"keyring.path":                      "keys",
	"log.caller":                        false,
	"log.colordisabled":                 false,
	"log.format":                        "text",
	"log.level":                         "info",
	"log.output":                        "stderr",
	"log.source":                        false,
	"log.stacktrace":                    false,
	"secretfile":                        ".env",
	"telemetry.disabled":                false,
	"datastore.nosigning":               false,
	"datastore.nosearchableencryption":  false,
	"datastore.defaultkeytype":          "secp256k1",
	"acp.document.type":                 "local",
	"replicator.retryintervals":         []int{30, 60, 120, 240, 480, 960, 1920},
}

// DefaultConfig returns a new config with default values.
func DefaultConfig() *viper.Viper {
	cfg := viper.New()

	cfg.AutomaticEnv()
	cfg.SetEnvPrefix("DEFRA")
	cfg.SetEnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_"))

	cfg.SetConfigName("config")
	cfg.SetConfigType("yaml")

	for key, val := range ConfigDefaults {
		cfg.SetDefault(key, val)
	}
	return cfg
}

// CreateConfig writes the default config file if one does not exist.
func CreateConfig(rootdir string, flags *pflag.FlagSet) error {
	cfg := DefaultConfig()
	cfg.AddConfigPath(rootdir)

	if err := bindConfigFlags(cfg, flags); err != nil {
		return err
	}
	// make sure rootdir exists
	if err := os.MkdirAll(rootdir, 0755); err != nil {
		return err
	}
	err := cfg.SafeWriteConfig()
	// error type is known and shouldn't be wrapped
	//
	//nolint:errorlint
	if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
		return nil
	}
	return err
}

// LoadConfig returns a new config with values from the config in the given rootdir.
func LoadConfig(rootdir string, flags *pflag.FlagSet) (*viper.Viper, error) {
	cfg := DefaultConfig()
	cfg.AddConfigPath(rootdir)

	// attempt to read the existing config
	err := cfg.ReadInConfig()
	// error type is known and shouldn't be wrapped
	//
	//nolint:errorlint
	if _, ok := err.(viper.ConfigFileNotFoundError); err != nil && !ok {
		return nil, err
	}
	// bind cli flags to config keys
	if err := bindConfigFlags(cfg, flags); err != nil {
		return nil, err
	}

	// make paths relative to the rootdir
	for _, key := range ConfigPaths {
		path := cfg.GetString(key)
		if path != "" && !filepath.IsAbs(path) {
			cfg.Set(key, filepath.Join(rootdir, path))
		}
	}

	// load environment variables from .env file if one exists
	err = godotenv.Load(cfg.GetString("secretfile"))
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		return nil, err
	}

	// set logging config
	corelog.SetConfig(corelog.Config{
		Level:            cfg.GetString("log.level"),
		Format:           cfg.GetString("log.format"),
		Output:           cfg.GetString("log.output"),
		EnableStackTrace: cfg.GetBool("log.stacktrace"),
		EnableSource:     cfg.GetBool("log.source"),
		DisableColor:     cfg.GetBool("log.colordisabled"),
	})

	// set logging config overrides
	corelog.SetConfigOverrides(cfg.GetString("log.overrides"))

	return cfg, nil
}

// bindConfigFlags binds the set of cli flags to config values.
func bindConfigFlags(cfg *viper.Viper, flags *pflag.FlagSet) error {
	var errs []error
	flags.VisitAll(func(f *pflag.Flag) {
		errs = append(errs, cfg.BindPFlag(ConfigFlags[f.Name], f))
	})
	return errors.Join(errs...)
}
